### DISCUSSION TITLE
Secure Boot: module signing with DKMS

### DISCUSSION THREAD

**Discussion by umlaeute** [umlaeute/v4l2loopback#563] [Q&A] [answered]
When secure boot is enabled, modules must be signed - otherwise the kernel will refuse to load them.
How?

**Comment by umlaeute** [comment, marked as answer]
For your convenience, the build system provides as `sign` target.

## Prerequisites
- a private key/public certificate pair for signing kernel modules
- the certificate must be known to the system 

How to create and enroll a signing certificate is outside the scope of this document.
For Debian systems (and probably others), check out the [documentation](https://wiki.debian.org/SecureBoot):
- [Creating a signing key](https://wiki.debian.org/SecureBoot#Generating_a_new_key)
- [Enrolling the signing key](https://wiki.debian.org/SecureBoot#Enrolling_your_key)



#### Simple usage

```sh
make sign
```

This assumes that the certificate/signing key are listed in `/etc/dkms/framework.conf` as  `mok_certificate` resp. `mok_signing_key`.
It also assumes that these files are readable by the current user and do not require a password to unlock.

#### Advanced usage

You can manually pass the signing key and certificate to the build process via the `KBUILD_SIGN_KEY` and `KBUILD_SIGN_CERT` variables.
f your key requires a passphrase, provide it via an (exported) envvar `KBUILD_SIGN_PIN`.

If the key (and/or certificate) require super-user proviliges to be read, run the build process through `sudo`, but make sure to preserve the `KBUILD_SIGN_PIN` environment variable.

```sh
read -s -p "Passphrase for your signing key: " KBUILD_SIGN_PIN
export KBUILD_SIGN_PIN
sudo --preserve-env=KBUILD_SIGN_PIN \
    make sign KBUILD_SIGN_CERT=/var/lib/shim-signed/mok/MOK.der KBUILD_SIGN_KEY=/var/lib/shim-signed/mok/MOK.priv
```

**Reply by I-NuKeS-I** [reply]
Hey man, could you help me install the module for a secure boot enabled fedora 42 KDE edition.

I've read the manual, but i cannot figure it out. please.

i was using your module no problems when i had secure boot disable, but recently i had to enable it for my dual boot, so i have to sign the module.

I tried the fedora repo (sudo dnf install v4l2loopback) but OBS does not like that one for some reason.

much appreciated
